pkg/utils/metadata.go

```diff
@@ -44,6 +53,22 @@ func ProjectID(project string, client metadataClient.Client) (string, error) {
 	return projectID, nil
 }
 
+// ProjectIDOrDefault returns the project ID by performing the following order:
+// 1) if the input project ID is valid, simply use it.
+// 2) if there is a PROJECT_ID environmental variable, use it.
+// 3) use metadataClient to resolve project id.
+func ProjectIDOrDefault(ctx context.Context, projectID string) (string, error) {
+	if projectID != "" {
+		return projectID, nil
+	}
+	// Parse project id from env var, if not found we simply continue
+	var env ProjectIDEnvConfig
+	if err := envconfig.Process("", &env); err != nil {
```

Comment:
It seems that we have two use cases: 1. read a PROJECT_ID from env var, or apply default; 2. get project from spec, or apply default. I recommend that the existing `ProjectID` function to be renamed to `ProjectIDOrDefault`, and rename this one `ProjectIDFromEnvOrDefault`.

Reply:
+1, Not sure whether sources need read PROJECT_ID from env var or not. This pr changes the original business logic.

Reply:
Sorry about the confusion here. This PR is not for clean up. It is mainly to introduce env var to sources #1781 , or in another word it is to have one single way of resolving project id in `knative-gcp`. @liu-cong That's why I used `ProjectIDOrDefault` instead of using implementation details as the name. The idea is that user shouldn't know the implementation details, or change their code when we decide to change the logic of resolving project id.

Reply:
Thanks for the explanation. Looks like we don't call the `ProjectID` any more in clients? If so, let's un-export it, or even inline the logic in the new function. I still think the `ProjectIDOrDefault` name does not reflect exactly what this function is doing. I think it should be renamed to either something verbose like `ProjectIDOrEnvVarOrDefault` or something general like `GetProjectID`

Reply:
We definitely want to avoid having `EnvVar` as we might change the logic in the future, eg we decided to drop support for env var and use config file instead. `GetProjectID` works to me too,  but I think it is pretty much same as `ProjectIDOrDefault`. As for the original `ProjectID`, I am not sure if we should remove it here, because it is still possible that the function can be used.

Reply:
Can you do a search to see if `ProjectID` is still used after this change? It will be confusing to have two helper functions that seem to be similar from their names. (To me, the old `ProjectID` does `ProjectIDOrDefault` logic)

Reply:
I did a search and replacement, so there won't be any usage of `ProjectID` now, but I am waiting for @capri-xiyue to confirm this is OK in sources. Either way the problem is really whether we want to keep such a helper function or not, if not, I think removing it is better.

Reply:
Remove the original `ProjectID` as suggested.